composer/datasets/in_context_learning_evaluation.py

```diff
@@ -752,134 +654,71 @@ class InContextLearningSchemaTaskDataset(InContextLearningMultipleChoiceTaskData
-                    context_options, continuation, gold_idx = self.samples[fewshot_idx][
-                        'context_options'], self.samples[fewshot_idx]['continuation'], self.samples[fewshot_idx]['gold']
-                    assert isinstance(gold_idx, int)
-                    context = context_options[gold_idx]
-                    if len(preamble) > 0:
-                        context = f'{example_delimiter}{context}'
-                    preamble += f'{context}{continuation_delimiter}{continuation}'
-
-            encoded_example = {}
-            context_options, continuation, gold_idx = self.samples[sample_idx]['context_options'], self.samples[
-                sample_idx]['continuation'], self.samples[sample_idx]['gold'],
-
-            # rstrip the continuation delimiter, because the prompt ending in a space results in degenerate output
-            continuation_delimiter_stripped = continuation_delimiter.rstrip()
-
-            if len(preamble) > 0:
-                context_options = [f'{example_delimiter}{c}{continuation_delimiter_stripped}' for c in context_options]
-            encoded_example['preamble'] = self.tokenizer(
-                preamble
-            )  # if the preamble is empty then these will be 0-length lists, unless the tokenizer adds special tokens to empty strings (e.g. OPT tokenizer)
-            if self.tokenizer.eos_token_id is not None and len(
-                    encoded_example['preamble']
-                ['input_ids']) > 1 and encoded_example['preamble']['input_ids'][-1] == self.tokenizer.eos_token_id:
-                encoded_example['preamble']['input_ids'] = encoded_example['preamble']['input_ids'][:-1]
-
-            encoded_example['gold_idx'] = gold_idx
-            encoded_example['context_options'] = [self.tokenizer(c, add_special_tokens=False) for c in context_options]
-
-            if self.prefix_space:
-                continuation = f' {continuation}' if not continuation.startswith(' ') else continuation
-            encoded_example['continuation'] = self.tokenizer(continuation, add_special_tokens=False)
-            examples.append(encoded_example)
-
-        return examples
+    def __init__(self, choices_key='context_options', *args, **kwargs):
+        super().__init__(choices_key=choices_key, *args, **kwargs)
+
+    def construct_context(self, sample, preceding_text: str = '', add_answer: bool = False):
+        context_options = sample['context_options']
+        gold_idx = sample['gold']
+        continuation = sample['continuation']
+        assert isinstance(gold_idx, int)
+        if add_answer:
+            context = context_options[gold_idx]
+            if len(preceding_text) > 0:
+                context = f'{self.example_delimiter}{context}'
+            context = f'{context}{self.continuation_delimiter}{continuation}'
+            return context
+        else:
+            # TODO: This is a kinda code-smelly bcus we return two different types
```

Comment:
Yeah could you explain why its like this?

Reply:
For schema tasks, the there are multiple contexts and a single answer. As such, when we return w/ an answer (such as in few_shot), we only need a single string. Else, we need a list. I'll redo this to make a new function for the non few-shot examples.

Reply:
I separated it into two different functions to circumvent this problem btw. New code reflects this, LMK if we think it's better.